File: src/api/sync.ts
import {
  DayReplayReference,
  RunSummary,
  RunSummaryEstablishment,
  RunSummaryUpdatedEstablishment,
  SyncMode,
  SyncRequestPayload,
  SyncRun,
  SyncState,
} from "../types";
import { request } from "./http";

export interface SyncRunResponse {
  id: string;
  scope_key: string;
  run_type: string;
  status: string;
  mode: SyncMode;
  replay_for_date: string | null;
  day_replay_reference: DayReplayReference;
  started_at: string;
  finished_at: string | null;
  api_call_count: number;
  google_api_call_count: number;
  fetched_records: number;
  created_records: number;
  updated_records: number;
  google_queue_count: number;
  google_eligible_count: number;
  google_matched_count: number;
  google_pending_count: number;
  google_immediate_matched_count: number;
  google_late_matched_count: number;
  last_cursor: string | null;
  query_checksum: string | null;
  resumed_from_run_id: string | null;
  notes: string | null;
  total_expected_records: number | null;
  progress: number | null;
  estimated_remaining_seconds: number | null;
  estimated_completion_at: string | null;
  summary: RunSummaryResponse | null;
  google_enabled: boolean;
  target_naf_codes: string[] | null;
  target_client_ids: string[] | null;
  notify_admins: boolean;
  day_replay_force_google: boolean;
}

export interface RunSummaryResponse {
  run: RunSummaryMetaResponse;
  stats: RunSummaryStatsResponse;
  samples: RunSummarySamplesResponse;
  email?: RunEmailSummaryResponse | null;
}

interface RunSummaryMetaResponse {
  id: string;
  scope_key: string;
  status: string;
  mode: SyncMode;
  started_at: string | null;
  finished_at: string | null;
  duration_seconds: number;
  page_count: number;
}

interface RunSummaryStatsResponse {
  mode: SyncMode;
  fetched_records: number;
  created_records: number;
  updated_records: number;
  api_call_count: number;
  google: RunSummaryGoogleStatsResponse;
  alerts: RunSummaryAlertsStatsResponse;
}

interface RunSummaryGoogleStatsResponse {
  enabled: boolean;
  api_call_count: number;
  queue_count: number;
  eligible_count: number;
  matched_count: number;
  immediate_matches: number;
  late_matches: number;
  pending_count: number;
}

interface RunSummaryAlertsStatsResponse {
  created: number;
  sent: number;
}

interface RunSummarySamplesResponse {
  new_establishments?: RunSummaryEstablishmentResponse[];
  updated_establishments?: RunSummaryUpdatedEstablishmentResponse[];
  google_late_matches?: RunSummaryEstablishmentResponse[];
  google_immediate_matches?: RunSummaryEstablishmentResponse[];
}

interface RunSummaryEstablishmentResponse {
  siret: string;
  name: string | null;
  code_postal: string | null;
  libelle_commune: string | null;
  naf_code: string | null;
  google_status: string | null;
  google_place_url: string | null;
  google_place_id: string | null;
  google_match_confidence: number | null;
  created_run_id: string | null;
  first_seen_at: string | null;
  last_seen_at: string | null;
}

interface RunSummaryUpdatedEstablishmentResponse extends RunSummaryEstablishmentResponse {
  changed_fields?: string[];
}

interface RunEmailSummaryResponse {
  sent: boolean;
  recipients?: string[];
  subject?: string | null;
  reason?: string | null;
}

export interface SyncStateResponse {
  scope_key: string;
  last_successful_run_id: string | null;
  last_cursor: string | null;
  cursor_completed: boolean;
  last_synced_at: string | null;
  last_total: number | null;
  last_treated_max: string | null;
  query_checksum: string | null;
  updated_at: string;
}

export interface DeleteRunResponse {
  establishments_deleted: number;
  alerts_deleted: number;
  states_reset: number;
  runs_updated: number;
  sync_run_deleted: boolean;
}

const mapRunSummaryEstablishment = (
  payload: RunSummaryEstablishmentResponse,
): RunSummaryEstablishment => ({
  siret: payload.siret,
  name: payload.name ?? null,
  codePostal: payload.code_postal ?? null,
  libelleCommune: payload.libelle_commune ?? null,
  nafCode: payload.naf_code ?? null,
  googleStatus: payload.google_status ?? null,
  googlePlaceUrl: payload.google_place_url ?? null,
  googlePlaceId: payload.google_place_id ?? null,
  googleMatchConfidence: payload.google_match_confidence ?? null,
  createdRunId: payload.created_run_id ?? null,
  firstSeenAt: payload.first_seen_at ?? null,
  lastSeenAt: payload.last_seen_at ?? null,
});

const mapRunSummaryUpdatedEstablishment = (
  payload: RunSummaryUpdatedEstablishmentResponse,
): RunSummaryUpdatedEstablishment => ({
  ...mapRunSummaryEstablishment(payload),
  changedFields: payload.changed_fields ?? [],
});

const mapRunSummary = (payload: RunSummaryResponse): RunSummary => ({
  run: {
    id: payload.run.id,
    scopeKey: payload.run.scope_key,
    status: payload.run.status,
    mode: payload.run.mode,
    startedAt: payload.run.started_at,
    finishedAt: payload.run.finished_at,
    durationSeconds: payload.run.duration_seconds,
    pageCount: payload.run.page_count,
  },
  stats: {
    mode: payload.stats.mode,
    fetchedRecords: payload.stats.fetched_records,
    createdRecords: payload.stats.created_records,
    updatedRecords: payload.stats.updated_records,
    apiCallCount: payload.stats.api_call_count,
    google: {
      enabled: payload.stats.google.enabled,
      apiCallCount: payload.stats.google.api_call_count,
      queueCount: payload.stats.google.queue_count,
      eligibleCount: payload.stats.google.eligible_count,
      matchedCount: payload.stats.google.matched_count,
      immediateMatches: payload.stats.google.immediate_matches,
      lateMatches: payload.stats.google.late_matches,
      pendingCount: payload.stats.google.pending_count,
    },
    alerts: {
      created: payload.stats.alerts.created,
      sent: payload.stats.alerts.sent,
    },
  },
  samples: {
    newEstablishments: (payload.samples.new_establishments ?? []).map(mapRunSummaryEstablishment),
    updatedEstablishments: (payload.samples.updated_establishments ?? []).map(
      mapRunSummaryUpdatedEstablishment,
    ),
    googleLateMatches: (payload.samples.google_late_matches ?? []).map(mapRunSummaryEstablishment),
    googleImmediateMatches: (payload.samples.google_immediate_matches ?? []).map(
      mapRunSummaryEstablishment,
    ),
  },
  email: payload.email
    ? {
        sent: payload.email.sent,
        recipients: payload.email.recipients ?? [],
        subject: payload.email.subject ?? null,
        reason: payload.email.reason ?? null,
      }
    : undefined,
});

export const mapSyncRun = (payload: SyncRunResponse): SyncRun => ({
  id: payload.id,
  scopeKey: payload.scope_key,
  runType: payload.run_type,
  status: payload.status,
  mode: payload.mode,
  replayForDate: payload.replay_for_date,
  startedAt: payload.started_at,
  finishedAt: payload.finished_at,
  apiCallCount: payload.api_call_count,
  googleApiCallCount: payload.google_api_call_count,
  fetchedRecords: payload.fetched_records,
  createdRecords: payload.created_records,
  updatedRecords: payload.updated_records,
  googleQueueCount: payload.google_queue_count,
  googleEligibleCount: payload.google_eligible_count,
  googleMatchedCount: payload.google_matched_count,
  googlePendingCount: payload.google_pending_count,
  googleImmediateMatchedCount: payload.google_immediate_matched_count,
  googleLateMatchedCount: payload.google_late_matched_count,
  lastCursor: payload.last_cursor,
  queryChecksum: payload.query_checksum,
  resumedFromRunId: payload.resumed_from_run_id,
  notes: payload.notes,
  totalExpectedRecords: payload.total_expected_records,
  progress: payload.progress,
  estimatedRemainingSeconds: payload.estimated_remaining_seconds,
  estimatedCompletionAt: payload.estimated_completion_at,
  googleEnabled: payload.google_enabled,
  targetNafCodes: payload.target_naf_codes ?? null,
  targetClientIds: payload.target_client_ids ?? null,
  notifyAdmins: payload.notify_admins,
  dayReplayForceGoogle: payload.day_replay_force_google,
  dayReplayReference: payload.day_replay_reference,
  summary: payload.summary ? mapRunSummary(payload.summary) : null,
});

export const mapSyncState = (payload: SyncStateResponse): SyncState => ({
  scopeKey: payload.scope_key,
  lastSuccessfulRunId: payload.last_successful_run_id,
  lastCursor: payload.last_cursor,
  cursorCompleted: payload.cursor_completed,
  lastSyncedAt: payload.last_synced_at,
  lastTotal: payload.last_total,
  lastTreatedMax: payload.last_treated_max,
  queryChecksum: payload.query_checksum,
  updatedAt: payload.updated_at,
});

const isSyncRunResponse = (payload: unknown): payload is SyncRunResponse => {
  return Boolean(payload) && typeof payload === "object" && "id" in (payload as Record<string, unknown>);
};

export interface TriggerSyncResult {
  run: SyncRun | null;
  status: number;
  detail?: string;
}

export const syncApi = {
  async fetchRuns(limit = 20): Promise<SyncRun[]> {
    const { data } = await request<SyncRunResponse[]>(`/admin/sync-runs?limit=${encodeURIComponent(limit)}`);
    return data.map(mapSyncRun);
  },

  async fetchState(): Promise<SyncState[]> {
    const { data } = await request<SyncStateResponse[]>("/admin/sync-state");
    return data.map(mapSyncState);
  },

  async trigger(payload: SyncRequestPayload): Promise<TriggerSyncResult> {
    const body: Record<string, unknown> = {};
    if (payload.checkForUpdates !== undefined) {
      body.check_for_updates = payload.checkForUpdates;
    }
    if (payload.mode) {
      body.mode = payload.mode;
    }
    if (payload.replayForDate) {
      body.replay_for_date = payload.replayForDate;
    }
    if (payload.replayReference) {
      body.replay_reference = payload.replayReference;
    }
    if (payload.nafCodes && payload.nafCodes.length > 0) {
      body.naf_codes = payload.nafCodes;
    }
    if (payload.targetClientIds && payload.targetClientIds.length > 0) {
      body.target_client_ids = payload.targetClientIds;
    }
    if (payload.notifyAdmins !== undefined) {
      body.notify_admins = payload.notifyAdmins;
    }
    if (payload.forceGoogleReplay !== undefined) {
      body.force_google_replay = payload.forceGoogleReplay;
    }
    const { data, status } = await request<SyncRunResponse | { detail?: string }>("/admin/sync", {
      method: "POST",
      body: JSON.stringify(body),
    });
    if (isSyncRunResponse(data)) {
      return { run: mapSyncRun(data), status };
    }
    const detail =
      typeof data === "object" && data !== null && "detail" in data
        ? String((data as { detail?: string }).detail)
        : undefined;
    return { run: null, status, detail };
  },

  async deleteRun(runId: string): Promise<DeleteRunResponse> {
    const { data } = await request<DeleteRunResponse>(`/admin/sync-runs/${encodeURIComponent(runId)}`, {
      method: "DELETE",
    });
    return data;
  },
};

File: app/api/schemas/sync.py
"""Schemas liés aux synchronisations (runs, états, rapports)."""
from __future__ import annotations

from datetime import datetime, date as Date
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, computed_field, field_validator, model_validator

from app.services.sync.mode import SyncMode
from app.services.sync.replay_reference import DayReplayReference, DEFAULT_DAY_REPLAY_REFERENCE
from app.utils.dates import utcnow
from app.utils.naf import normalize_naf_code


class SyncRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    scope_key: str
    run_type: str
    status: str
    mode: SyncMode
    replay_for_date: Date | None = Field(default=None)
    started_at: datetime
    finished_at: datetime | None
    api_call_count: int
    google_api_call_count: int
    fetched_records: int
    created_records: int
    google_queue_count: int
    google_eligible_count: int
    google_matched_count: int
    google_pending_count: int
    google_immediate_matched_count: int
    google_late_matched_count: int
    # LinkedIn enrichment progress
    linkedin_queue_count: int = 0
    linkedin_searched_count: int = 0
    linkedin_found_count: int = 0
    linkedin_not_found_count: int = 0
    linkedin_error_count: int = 0
    # Website scraping progress
    website_scrape_count: int = 0
    website_scrape_success_count: int = 0
    updated_records: int
    summary: dict[str, Any] | None = None
    last_cursor: str | None
    query_checksum: str | None
    resumed_from_run_id: UUID | None
    notes: str | None
    target_naf_codes: list[str] | None = Field(
        default=None,
        description="Liste optionnelle des codes NAF ciblés par ce run.",
    )
    target_client_ids: list[UUID] | None = Field(
        default=None,
        description="Clients explicitement ciblés lors de ce run (mode 'day_replay' uniquement).",
    )
    notify_admins: bool = Field(
        default=True,
        description="Indique si les alertes de ce run sont envoyées aux administrateurs.",
    )
    day_replay_force_google: bool = Field(
        default=False,
        description="Indique si les appels Google ont été forcés lors d'un rejeu.",
    )
    day_replay_reference: DayReplayReference = Field(
        default=DEFAULT_DAY_REPLAY_REFERENCE,
        description="Source de vérité utilisée pour rejouer une journée (date de création ou d'insertion).",
    )
    months_back: int | None = Field(
        default=None,
        description="Nombre de mois dans le passé à inclure dans la collecte (None = jour courant uniquement).",
    )
    total_expected_records: int | None = None
    progress: float | None = None
    estimated_remaining_seconds: float | None = None
    estimated_completion_at: datetime | None = None

    @computed_field
    @property
    def google_enabled(self) -> bool:
        return self.mode != SyncMode.SIRENE_ONLY

    @computed_field
    @property
    def linkedin_enabled(self) -> bool:
        return self.mode in {SyncMode.FULL, SyncMode.LINKEDIN_PENDING, SyncMode.LINKEDIN_REFRESH}


class SyncStateOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    scope_key: str
    last_successful_run_id: UUID | None
    last_cursor: str | None
    cursor_completed: bool
    last_synced_at: datetime | None
    last_total: int | None
    last_treated_max: datetime | None
    last_creation_date: Date | None
    query_checksum: str | None
    updated_at: datetime


class RunEstablishmentSummary(BaseModel):
    siret: str
    name: str | None = None
    code_postal: str | None = None
    libelle_commune: str | None = None
    google_status: str | None = None
    google_place_url: str | None = None
    google_place_id: str | None = None
    created_run_id: UUID | None = None
    first_seen_at: datetime | None = None


class RunUpdatedEstablishmentSummary(RunEstablishmentSummary):
    changed_fields: list[str] = Field(default_factory=list)


class RunSummaryStats(BaseModel):
    new_establishments: int
    updated_establishments: int
    fetched_records: int
    api_call_count: int
    google_total_matches: int
    google_immediate_matches: int
    google_late_matches: int
    google_api_call_count: int
    alerts_created: int
    alerts_sent: int
    page_count: int
    duration_seconds: float


class RunEmailSummary(BaseModel):
    sent: bool
    recipients: list[str] = Field(default_factory=list)
    subject: str | None = None
    reason: str | None = None


class SyncRunReport(BaseModel):
    run: SyncRunOut
    stats: RunSummaryStats
    new_establishments: list[RunEstablishmentSummary]
    updated_establishments: list[RunUpdatedEstablishmentSummary]
    google_immediate_matches: list[RunEstablishmentSummary]
    google_late_matches: list[RunEstablishmentSummary]
    email: RunEmailSummary | None = None


class SyncRequest(BaseModel):
    check_for_updates: bool = Field(
        default=False,
        description="Vérifie le service informations Sirene et annule si aucune mise à jour n'est disponible.",
    )
    mode: SyncMode = Field(
        default=SyncMode.FULL,
        description=(
            "Mode d'exécution: 'full' exécute l'enrichissement Google et LinkedIn, 'sirene_only' le désactive, "
            "'google_refresh' relance Google sur les établissements filtrés par statuts, "
            "'linkedin_pending' relance uniquement les dirigeants jamais enrichis par LinkedIn, "
            "'linkedin_refresh' relance une détection LinkedIn sur tous les dirigeants, "
            "'day_replay' rejoue une journée complète en limitant les alertes aux administrateurs, "
            "'website_scrape' scrape les sites web des fiches Google."
        ),
    )
    reset_google_state: bool = Field(
        default=False,
        description=(
            "Disponible uniquement en mode 'google_refresh' (forcé à true) : réinitialise les données Google "
            "avant relance (reset biz-by-biz)."
        ),
    )
    replay_for_date: Date | None = Field(
        default=None,
        description="Limite la collecte aux établissements créés à la date indiquée (mode 'day_replay' uniquement).",
    )
    naf_codes: list[str] | None = Field(
        default=None,
        description="Filtre optionnel de codes NAF (ex: 5610A) à rejouer quel que soit le mode.",
    )
    target_client_ids: list[UUID] | None = Field(
        default=None,
        description="Limiter l'envoi des alertes à une liste précise de clients (mode 'day_replay').",
    )
    linkedin_statuses: list[str] | None = Field(
        default=None,
        description=(
            "Statuts LinkedIn à relancer lors d'un run LinkedIn-only (pending, found, not_found, error, insufficient)."
        ),
    )
    google_statuses: list[str] | None = Field(
        default=None,
        description=(
            "Statuts Google à relancer lors d'un run Google-only (ex: pending, found, not_found, insufficient, type_mismatch)."
        ),
    )
    website_statuses: list[str] | None = Field(
        default=None,
        description=(
            "Statuts de scraping site web à cibler (scraped, not_scraped). "
            "Mode 'website_scrape' uniquement."
        ),
    )
    notify_admins: bool = Field(
        default=True,
        description="Autorise l'envoi des alertes administrateurs lors d'un rejeu (mode 'day_replay').",
    )
    force_google_replay: bool = Field(
        default=False,
        description="Force les appels Google lors d'un rejeu même si des fiches existent déjà.",
    )
    replay_reference: DayReplayReference = Field(
        default=DEFAULT_DAY_REPLAY_REFERENCE,
        description=(
            "Contrôle le critère de rejeu: 'creation_date' récupère les établissements par date de création, "
            "'insertion_date' par date d'insertion en base (mode 'day_replay')."
        ),
    )
    months_back: int | None = Field(
        default=None,
        ge=1,
        le=24,
        description=(
            "Nombre de mois dans le passé à inclure dans la collecte Sirene. "
            "Si non spécifié, seul le jour courant est synchronisé (synchro incrémentale standard)."
        ),
    )

    @field_validator("linkedin_statuses")
    @classmethod
    def validate_linkedin_statuses(cls, value: list[str] | None) -> list[str] | None:
        if not value:
            return None
        allowed = {"pending", "found", "not_found", "error", "insufficient"}
        normalized: list[str] = []
        seen: set[str] = set()
        for raw in value:
            candidate = (raw or "").strip().lower()
            if not candidate:
                continue
            if candidate not in allowed:
                raise ValueError("Statuts LinkedIn acceptés: pending, found, not_found, error, insufficient.")
            if candidate in seen:
                continue
            seen.add(candidate)
            normalized.append(candidate)
        return normalized or None

    @field_validator("google_statuses")
    @classmethod
    def validate_google_statuses(cls, value: list[str] | None) -> list[str] | None:
        if not value:
            return None
        normalized: list[str] = []
        seen: set[str] = set()
        for raw in value:
            candidate = (raw or "").strip().lower()
            if not candidate:
                continue
            if candidate in seen:
                continue
            seen.add(candidate)
            normalized.append(candidate)
        return normalized or None

    @field_validator("website_statuses")
    @classmethod
    def validate_website_statuses(cls, value: list[str] | None) -> list[str] | None:
        if not value:
            return None
        allowed = {"scraped", "not_scraped"}
        normalized: list[str] = []
        seen: set[str] = set()
        for raw in value:
            candidate = (raw or "").strip().lower()
            if not candidate:
                continue
            if candidate not in allowed:
                raise ValueError("Statuts site web acceptés: scraped, not_scraped.")
            if candidate in seen:
                continue
            seen.add(candidate)
            normalized.append(candidate)
        return normalized or None

    @field_validator("naf_codes")
    @classmethod
    def validate_naf_codes(cls, value: list[str] | None) -> list[str] | None:
        if not value:
            return None
        normalized: list[str] = []
        seen: set[str] = set()
        for raw in value:
            candidate = (raw or "").strip().upper().replace(" ", "")
            normalized_code = normalize_naf_code(candidate)
            if not normalized_code:
                raise ValueError("Chaque code NAF doit contenir 4 chiffres suivis d'une lettre (ex: 56.10A).")
            if normalized_code in seen:
                continue
            seen.add(normalized_code)
            normalized.append(normalized_code)
            if len(normalized) > 25:
                raise ValueError("Maximum 25 codes NAF ciblés par synchronisation.")
        return normalized or None

    @field_validator("target_client_ids")
    @classmethod
    def validate_target_clients(cls, value: list[UUID] | None) -> list[UUID] | None:
        if not value:
            return None
        normalized: list[UUID] = []
        seen: set[str] = set()
        for raw in value:
            try:
                candidate = UUID(str(raw))
            except (TypeError, ValueError) as exc:  # noqa: PERF203 - explicit error for clarity
                raise ValueError("Chaque client ciblé doit être un UUID valide.") from exc
            candidate_key = str(candidate)
            if candidate_key in seen:
                continue
            seen.add(candidate_key)
            normalized.append(candidate)
            if len(normalized) > 50:
                raise ValueError("Maximum 50 clients ciblés par synchronisation.")
        return normalized or None

    @model_validator(mode="after")
    def validate_replay_settings(self) -> "SyncRequest":
        if "reset_google_state" in self.model_fields_set and self.mode is not SyncMode.GOOGLE_REFRESH:
            raise ValueError("Le paramètre reset_google_state est disponible uniquement pour les modes Google-only.")
        if self.mode is SyncMode.GOOGLE_REFRESH and not self.google_statuses:
            raise ValueError("Merci de sélectionner au moins un statut Google à relancer.")
        if self.google_statuses and self.mode is not SyncMode.GOOGLE_REFRESH:
            raise ValueError("Les statuts Google sont disponibles uniquement en mode 'google_refresh'.")
        if self.mode is SyncMode.GOOGLE_REFRESH:
            self.reset_google_state = True
        if self.mode.requires_replay_date and not self.replay_for_date:
            raise ValueError("Une date est requise pour rejouer une journée.")
        if self.replay_for_date and not self.mode.requires_replay_date:
            raise ValueError("La date de rejeu n'est disponible qu'en mode 'day_replay'.")
        if self.replay_for_date and self.replay_for_date > utcnow().date():
            raise ValueError("Impossible de rejouer une journée future.")
        if self.target_client_ids and self.mode is not SyncMode.DAY_REPLAY:
            raise ValueError("Le ciblage de clients est disponible uniquement en mode 'day_replay'.")
        if self.notify_admins is False and self.mode is not SyncMode.DAY_REPLAY:
            raise ValueError("La désactivation des notifications admin est réservée au mode 'day_replay'.")
        if self.force_google_replay and self.mode is not SyncMode.DAY_REPLAY:
            raise ValueError("Le forçage des appels Google est disponible uniquement en mode 'day_replay'.")
        if self.replay_reference is not DEFAULT_DAY_REPLAY_REFERENCE and self.mode is not SyncMode.DAY_REPLAY:
            raise ValueError("Le choix de la référence de rejeu est réservé au mode 'day_replay'.")
        return self


class DeleteRunResult(BaseModel):
    establishments_deleted: int = Field(description="Nombre d'établissements supprimés.")
    alerts_deleted: int = Field(description="Nombre d'alertes supprimées.")
    states_reset: int = Field(description="Nombre d'états de synchronisation remis à zéro.")
    runs_updated: int = Field(description="Nombre de runs mis à jour (liens de reprise supprimés).")
    sync_run_deleted: bool = Field(description="Indique si le run a été supprimé.")


__all__ = [
    "DeleteRunResult",
    "RunEmailSummary",
    "RunEstablishmentSummary",
    "RunSummaryStats",
    "RunUpdatedEstablishmentSummary",
    "SyncRequest",
    "SyncRunOut",
    "SyncRunReport",
    "SyncStateOut",
]
